docs/DesignDocs/CustomAttestation_V3.md

```diff
@@ -0,0 +1,423 @@
+
+This document describes the added and affected APIs as well as keeping the existing support for version 2 of the API. This document assumes the user is familiar with the SGX attestation and [custom attestation](https://github.com/openenclave/openenclave/blob/master/docs/DesignDocs/CustomAttestation.md) with plugins.
+
+Motivation
+----------
+
+New APIs were added to help customers who would like to extend Open Enclave's current attestation structure with extra information while keep TEE-specific information within attester and verifier plugins.
+
+1. The application can query built-in SGX attesters and verifiers to be registered. This allows the plugin library to return attesters and verifiers based on configuration.
+
+```C
+   // Old
+   oe_register_attester(oe_sgx_attesterX(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterY(), NULL, 0);
+   oe_register_attester(oe_sgx_attesterZ(), NULL, 0);
+
+   // New
+   oe_sgx_get_attesters(&attesters, &attesters_length);
+   for (n = 0; n < attesters_length; n++)
+      oe_register_attester(&attesters[n], NULL, 0);
+```
+
+2. The attester application can rely on the SDK to select a matching evidence format.
+
+```C
+   // Old
+   const oe_uuid_t attester_formats[3] = {{ATTESTERX_UUID}, {ATTESTERY_UUID}, {ATTESTERZ_UUID}};
+   selected_format = NULL;
+   for (n = 0; n < requested_formats_length && !selected_format; n++)
+   {
+       for (m = 0; m < countof(attester_formats) && !selected_format; m++)
+       {
+           if (requested_formats[n] == attester_formats[m])
+               selected_format = &attester_formats[m];
+       }
+   }
+
+   // New
+   oe_select_attester_format(
+       requested_formats,
+       requested_formats_length,
+       &selected_format);
+```
+
+3. The application gets optional evidence parameters from verifiers and stays TEE-agnostic when generating evidences.
+
+```C
+   // Old
+   uint8_t opt_paramsX[] = { OPT_PARAMS_X };
+   uint8_t opt_paramsY[] = { OPT_PARAMS_Y };
+   if (format == ATTESTERX_UUID)
+   {
+      oe_get_evidence(&format, 0, NULL, 0, opt_paramsX, sizeof(opt_paramsX), &evidence, &evidence_size, NULL, NULL);
+   }
+   else if (format == ATTESTERY_UUID)
+   {
+       oe_get_evidence(&format, 0, NULL, 0, opt_paramsY, sizeof(opt_paramsY), &evidence, &evidence_size, NULL, NULL);
+   }
+
+   // New
+   uint8_t* opt_params;
+   size_t opt_params_size;
+   oe_get_evidence_parameters(&format, &opt_params, &opt_params_size);
+
+   oe_get_evidence(&format, 0, NULL, 0, opt_params, opt_params_size, &evidence, &evidence_size, NULL, NULL);
+```
+
+4. SGX report (referred as local attestation in Open Enclave) and SGX quote (referred as remote attestation in Open Enclave) generation takes different inputs and generates data of different structures. The local/remote attestation concept is SGX specific. To prevent confusion at evidence generation, the OE_EVIDENCE_FLAGS_LOCAL_ATTESTATION and OE_EVIDENCE_FLAGS_REMOTE_ATTESTATION flags were deprecated.
```

Comment:
Grammar is not quite correct.  "SGX report" is not "referred as" local attestation.   Maybe "SGX report (an SGX-specific term for local attestation in Open Enclave)"?